clang/runtime/dpct-rt/include/sparse_utils.hpp

```diff
@@ -1164,7 +1195,228 @@ inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
+  T *new_val_b = (T *)val_b;
+  if (is_estimation) {
+    new_val_a = (T *)dpct::dpct_malloc(nnz_a * sizeof(T));
+    new_val_b = (T *)dpct::dpct_malloc(nnz_b * sizeof(T));
+    float one_f = 1.f;
+    unsigned int one_ui = *reinterpret_cast<unsigned int *>(&one_f);
+    dpct::dpct_memset_d32(new_val_a, one_ui, nnz_a);
+    dpct::dpct_memset_d32(new_val_b, one_ui, nnz_b);
+  }
+
+  int rows_c = m;
+  int cols_c = n;
+  int rows_a = (trans_a == oneapi::mkl::transpose::nontrans) ? m : k;
+  int cols_a = (trans_a == oneapi::mkl::transpose::nontrans) ? k : m;
+  int rows_b = (trans_b == oneapi::mkl::transpose::nontrans) ? k : n;
+  int cols_b = (trans_b == oneapi::mkl::transpose::nontrans) ? n : k;
+
+  auto data_row_ptr_a = dpct::detail::get_memory<int>(row_ptr_a);
+  auto data_col_ind_a = dpct::detail::get_memory<int>(col_ind_a);
+  auto data_val_a = dpct::detail::get_memory<Ty>(new_val_a);
+  oneapi::mkl::sparse::set_csr_data(queue, matrix_handle_a, rows_a, cols_a,
+                                    info_a->get_index_base(), data_row_ptr_a,
+                                    data_col_ind_a, data_val_a);
+  auto data_row_ptr_b = dpct::detail::get_memory<int>(row_ptr_b);
+  auto data_col_ind_b = dpct::detail::get_memory<int>(col_ind_b);
+  auto data_val_b = dpct::detail::get_memory<Ty>(new_val_b);
+  oneapi::mkl::sparse::set_csr_data(queue, matrix_handle_b, rows_b, cols_b,
+                                    info_b->get_index_base(), data_row_ptr_b,
+                                    data_col_ind_b, data_val_b);
+
+  auto data_row_ptr_c = dpct::detail::get_memory<int>(row_ptr_c);
+  auto data_col_ind_c = dpct::detail::get_memory<int>(nullptr);
+  auto data_val_c = dpct::detail::get_memory<Ty>(nullptr);
+  oneapi::mkl::sparse::set_csr_data(queue, matrix_handle_c, rows_c, cols_c,
+                                    info_c->get_index_base(), data_row_ptr_c,
+                                    data_col_ind_c, data_val_c);
+
+  oneapi::mkl::sparse::matmat_descr_t matmat_descr = nullptr;
+  oneapi::mkl::sparse::init_matmat_descr(&matmat_descr);
+  oneapi::mkl::sparse::set_matmat_data(
+      matmat_descr, oneapi::mkl::sparse::matrix_view_descr::general, trans_a,
+      oneapi::mkl::sparse::matrix_view_descr::general, trans_b,
+      oneapi::mkl::sparse::matrix_view_descr::general);
+
+  oneapi::mkl::sparse::matmat(
+      queue, matrix_handle_a, matrix_handle_b, matrix_handle_c,
+      oneapi::mkl::sparse::matmat_request::work_estimation, matmat_descr,
+      nullptr, nullptr
+#ifndef DPCT_USM_LEVEL_NONE
+      ,
+      {}
+#endif
+  );
+  queue.wait();
+
+  oneapi::mkl::sparse::matmat(queue, matrix_handle_a, matrix_handle_b,
+                              matrix_handle_c,
+                              oneapi::mkl::sparse::matmat_request::compute,
+                              matmat_descr, nullptr, nullptr
+#ifndef DPCT_USM_LEVEL_NONE
+                              ,
+                              {}
 #endif
+  );
+
+  if (is_estimation) {
```

Comment:
so first time through would call 

is_estimation:  work_estimation + compute + get_nnz

second time through would call 

work_estimation + compute + finalize ??

Isn't this a little redundant and doing way more work than is necessary?  is csrgemm_nnz_estimate optional ?  or where does C matrix actually get allocated ? 

Can you actaully walk me through a use case of how it gets mapped ?

Reply:
I would like to talk through this before you merge it.  I came back after the initial review and approval and found some questions still in my mind that I hadn't noticed before I approved.  :)

Reply:
Hi @spencerpatty , the example is:
<img width="917" alt="image" src="https://github.com/oneapi-src/SYCLomatic/assets/83437071/40604127-e859-4f1e-a0eb-7fd516f2c4f7">

Reply:
And the full example files are:
[example.zip](https://github.com/oneapi-src/SYCLomatic/files/13507524/example.zip)